apps/server/src/shared/domain/entity/task.entity.ts

```diff
@@ -99,102 +104,153 @@ export class Task extends BaseEntityWithTimestamps implements ILearnroomElement,
+
+		const isCourseFinished = this.course ? this.course.isFinished() : false;
+
+		const isFinishedForUser = isUserInFinishedUser || isCourseFinished;
+
+		return isFinishedForUser;
+	}
+
+	public isDraft(): boolean {
 		// private can be undefined in the database
 		return !!this.private;
 	}
 
-	isPublished(): boolean {
+	public isPublished(): boolean {
 		if (this.isDraft()) {
 			return false;
 		}
 		if (this.availableDate && this.availableDate > new Date(Date.now())) {
 			return false;
 		}
+
 		return true;
 	}
 
-	isPlanned(): boolean {
+	public isPlanned(): boolean {
 		if (this.isDraft()) {
 			return false;
 		}
 		if (this.availableDate && this.availableDate > new Date(Date.now())) {
 			return true;
 		}
+
 		return false;
 	}
 
-	areSubmissionsPublic(): boolean {
-		return !!this.publicSubmissions;
+	private getSubmittedSubmissions(): Submission[] {
+		const submissions = this.getSubmissionItems();
+		const gradedSubmissions = submissions.filter((submission) => submission.isSubmitted());
+
+		return gradedSubmissions;
 	}
 
-	private getSubmissionItems(): Submission[] {
-		if (!this.submissions.isInitialized(true)) {
-			throw new Error('Submissions items are not loaded.');
-		}
-		const submissions = this.submissions.getItems();
-		return submissions;
+	public areSubmissionsPublic(): boolean {
+		const areSubmissionsPublic = !!this.publicSubmissions;
+
+		return areSubmissionsPublic;
 	}
 
-	getSubmittedUserIds(): EntityId[] {
-		const submittedUserIds = this.getSubmissionItems().map((submission) => submission.student.id);
-		const uniqueSubmittedUserIds = [...new Set(submittedUserIds)];
+	private getGradedSubmissions(): Submission[] {
+		const submissions = this.getSubmissionItems();
+		const gradedSubmissions = submissions.filter((submission) => submission.isGraded());
 
-		return uniqueSubmittedUserIds;
+		return gradedSubmissions;
 	}
 
-	getNumberOfSubmittedUsers(): number {
-		const submittedUserIds = this.getSubmittedUserIds();
-		const count = submittedUserIds.length;
+	private isSubmittedForUser(user: User): boolean {
+		const submissions = this.getSubmissionItems();
+		const isSubmitted = submissions.some((submission) => submission.isSubmittedForUser(user));
 
-		return count;
+		return isSubmitted;
 	}
 
-	getGradedUserIds(): EntityId[] {
-		const gradedUserIds = this.getSubmissionItems()
-			.filter((submission) => submission.isGraded())
-			.map((submission) => submission.student.id);
-		const uniqueGradedUserIds = [...new Set(gradedUserIds)];
+	private isGradedForUser(user: User): boolean {
+		const submissions = this.getSubmissionItems();
+		const graded = submissions.some((submission) => submission.isGradedForUser(user));
 
-		return uniqueGradedUserIds;
+		return graded;
 	}
 
-	getNumberOfGradedUsers(): number {
-		const gradedUserIds = this.getGradedUserIds();
-		const count = gradedUserIds.length;
+	private getUniqueSubmitterIds(): EntityId[] {
```

Comment:
It must be enforced because it is used for numberOfSubmitter but i change the output and name of the method to the real need. numberOfSubmitter(): number same for getUniqueSubmitterWithGrade.